src/contract/link.rs

```diff
@@ -0,0 +1,233 @@
+//! This module implements typed linking for contracts.
+
+use crate::contract::deploy::Deploy;
+use crate::errors::LinkerError;
+use ethcontract_common::abi::ErrorKind as AbiErrorKind;
+use ethcontract_common::Bytecode;
+use std::collections::{HashMap};
+use std::marker::PhantomData;
+use web3::api::Web3;
+use web3::contract::tokens::Tokenize;
+use web3::types::Address;
+use web3::Transport;
+
+/// A trait that is implemented by a library instance and can used for linking.
+pub trait LibraryInstance {
+    /// The name of the library.
+    fn name(&self) -> &str;
+
+    /// The address of the library.
+    fn address(&self) -> Address;
+}
+
+impl<T> LibraryInstance for &'_ T
+where
+    T: LibraryInstance,
+{
+    #[inline(always)]
+    fn name(&self) -> &str {
+        <T as LibraryInstance>::name(self)
+    }
+
+    #[inline(always)]
+    fn address(&self) -> Address {
+        <T as LibraryInstance>::address(self)
+    }
+}
+
+/// A trait that is implemented by a library type that can be deployed alongside
+/// a contract when deploying the contract with its libraries.
+pub trait DeployLibrary {
```

Comment:
nit: DeployableLibrary?

Reply:
Sure, none of these names are final. I'm still figuring out the appropriate abstractions to use.